[WTF] Define WTF_MAKE_CONFIGURABLE_ALLOCATED for container types

Achierius opened the pull request:
#### f77dccdb4728f0595aa5189fcf9a59e02494f81a
<pre>
[WTF] Define WTF_MAKE_CONFIGURABLE_ALLOCATED for container types
<a href="https://bugs.webkit.org/show_bug.cgi?id=286711">https://bugs.webkit.org/show_bug.cgi?id=286711</a>
<a href="https://rdar.apple.com/143849704">rdar://143849704</a>

Reviewed by Yusuke Suzuki.

This allows us to specify that these types are allocated using the same
allocator as provided for their backing store, which is a more
comfortable default than &quot;always use FastMalloc no matter what the
caller tells us they want&quot;.

If in the future someone does want to use separate allocators for the
two, we can add another template parameter to configure the two
separately as needed.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/Bag.h:
* Source/WTF/wtf/CheckedRef.h:
(WTF::CanMakeCheckedPtr::~CanMakeCheckedPtr):
(WTF::CanMakeThreadSafeCheckedPtr::~CanMakeThreadSafeCheckedPtr):
* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/FastMalloc.h:
(WTF::FastMalloc::fastFree):
(WTF::FastCompactMalloc::fastFree):
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope): Deleted.
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope): Deleted.
(WTF::TryMallocReturnValue::TryMallocReturnValue): Deleted.
(WTF::TryMallocReturnValue::~TryMallocReturnValue): Deleted.
(WTF::TryMallocReturnValue::getValue): Deleted.
* Source/WTF/wtf/FixedVector.h:
* Source/WTF/wtf/MallocCommon.cpp: Added.
(WTF::ForbidMallocUseForCurrentThreadScope::ForbidMallocUseForCurrentThreadScope):
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope):
(WTF::DisableMallocRestrictionsForCurrentThreadScope::DisableMallocRestrictionsForCurrentThreadScope):
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope):
* Source/WTF/wtf/MallocCommon.h: Added.
(WTF::TryMallocReturnValue::TryMallocReturnValue):
(WTF::TryMallocReturnValue::~TryMallocReturnValue):
(WTF::TryMallocReturnValue::getValue):
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope):
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope):
* Source/WTF/wtf/MallocSpan.h:
* Source/WTF/wtf/SegmentedVector.h:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::makeUnique):
(WTF::makeUniqueWithoutRefCountedCheck):
* Source/WTF/wtf/UniqueRef.h:
(WTF::makeUniqueRefWithoutRefCountedCheck):
(WTF::makeUniqueRef):
* Source/WTF/wtf/Vector.h:
* Source/WTF/wtf/text/WTFString.h:
* Source/bmalloc/bmalloc/IsoHeap.h:
* Source/bmalloc/bmalloc/IsoHeapInlines.h:
* Source/bmalloc/bmalloc/TZoneHeap.h:
* Source/bmalloc/bmalloc/TZoneHeapInlines.h:

Canonical link: <a href="https://commits.webkit.org/290626@main">https://commits.webkit.org/290626@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a0201b5cf4e5b73e48255f7f99e37c5be7d2d63d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90522 "8 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10054 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45458 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95533 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41304 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10447 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18373 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69667 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27227 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93523 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7974 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82095 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50015 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7697 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36462 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40434 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83328 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78029 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37534 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97424 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89302 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17714 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13009 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78730 "layout-tests (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17972 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77919 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77930 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22321 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20943 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14267 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17724 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23056 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/111791 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17463 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20918 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19247 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/WTF/wtf/MallocCommon.h:
```diff
@@ -0,0 +1,246 @@
+    using WTFIs ## family ## Allocated  = int; \
+
+#if ENABLE(MALLOC_HEAP_BREAKDOWN)
+#define WTF_MAKE_CONFIGURABLE_ALLOCATED_WITH_HEAP_IDENTIFIER_IMPL(classname, alloc, family) \
```
Ditto, for `WTF_MAKE_FAST_ALLOCATED_WITH_HEAP_IDENTIFIER_IMPL`.

Achierius replied:
Discussed elsewhere, will do in another PR + add a FIXME here

hyjorc1 commented on Source/WTF/wtf/MallocCommon.h:
```diff
@@ -0,0 +1,246 @@
+using WTF::DisableMallocRestrictionsForCurrentThreadScope;
+using WTF::ForbidMallocUseForCurrentThreadScope;
+
+#define WTF_MAKE_CONFIGURABLE_ALLOCATED_IMPL(alloc, family) \
```
The logic seems identical to `WTF_MAKE_FAST_ALLOCATED_IMPL`. Can we refactor them into a single macro?

Achierius replied:
Discussed elsewhere, will do in another PR + add a FIXME here

hyjorc1 commented on Source/WTF/wtf/SegmentedVector.h:
```diff
@@ -83,7 +84,7 @@ namespace WTF {
         }
 
-        SegmentedVector<T, SegmentSize>& m_vector;
+        SegmentedVector<T, SegmentSize, Malloc>& m_vector;
```
Oh, this bug was introduced in https://commits.webkit.org/289249@main. Hmm... could you please review the previous similar changes to ensure we haven’t missed any template arguments?

Achierius replied:
Yep -- this was introduced when I was pulling pieces off of my much-larger diff (git add -p) and I missed this segment. I've more thoroughly vetted this patch by making sure it works with the sequestered-arena patch enabled (as that's the only thing which changes these to use a new allocator) + did another review of the other types I touched in the previous two patches, so I think it should be good now.

hyjorc1 commented on Source/WTF/wtf/MallocSpan.h:
```diff
@@ -37,6 +38,7 @@ namespace WTF {
 
 template<typename T, typename Malloc = FastMalloc> class MallocSpan {
     WTF_MAKE_NONCOPYABLE(MallocSpan);
+    WTF_MAKE_CONFIGURABLE_ALLOCATED(Malloc);
```
It looks like `WTF_MAKE_CONFIGURABLE_ALLOCATED(Malloc)` is unnecessary here. `MallocSpan` itself is never dynamically allocated, and Malloc is only used for managing the span’s buffer, not MallocSpan instances. Should we remove this to avoid unnecessary complexity?

Achierius replied:
Fixed

hyjorc1 commented on Source/WTF/wtf/FixedVector.h:
```diff
@@ -26,12 +26,15 @@
 
 template<typename T, typename Malloc>
 class FixedVector {
+    WTF_MAKE_CONFIGURABLE_ALLOCATED(Malloc);
```
It looks like `WTF_MAKE_CONFIGURABLE_ALLOCATED(Malloc)` is unnecessary here. `FixedVector<T, Malloc>` itself is never dynamically allocated, and Malloc is only used for managing the underlying storage, not the FixedVector instance. Should we remove this to avoid unnecessary complexity?

Achierius replied:
I'm not convinced it won't be in the future, though -- SegmentedVector, for example, is quite similar but is dynamically allocated, and if someone decides to do the same for FixedVector I don't think we can assume that they'll remember to make it arena-allocated.

Constellation commented on Source/WTF/wtf/MallocCommon.h:
```diff
@@ -0,0 +1,248 @@
+    return data;
+}
+
+class ForbidMallocUseForCurrentThreadScope {
```
You can use `WTF_MAKE_NONCOPYABLE` and `WTF_MAKE_NONMOVABLE` (`Noncopyable.h` and `Nonmovable.h`)

Achierius replied:
This is an existing piece of code -- I just moved it from FastMalloc.h to here so that it can also apply to the new arena allocator. It's also different from WTF_MAKE_NONCOPYABLE/MOVABLE since it prevents allocation of *all* types, but for the thread you're in -- whereas those macros forbid it for a specific type.

E.g. we use it in a few places in the audio stack where I assume they want to make sure we don't accidentally allocate and thus introduce jitter.

Constellation commented on Source/WTF/wtf/MallocCommon.h:
```diff
@@ -0,0 +1,248 @@
+    ~DisableMallocRestrictionsForCurrentThreadScope() { }
+#endif
+
+    DisableMallocRestrictionsForCurrentThreadScope(const DisableMallocRestrictionsForCurrentThreadScope&) = delete;
+    DisableMallocRestrictionsForCurrentThreadScope(DisableMallocRestrictionsForCurrentThreadScope&&) = delete;
+    DisableMallocRestrictionsForCurrentThreadScope& operator=(const DisableMallocRestrictionsForCurrentThreadScope&) = delete;
+    DisableMallocRestrictionsForCurrentThreadScope& operator=(DisableMallocRestrictionsForCurrentThreadScope&&) = delete;
```
Ditto.

Achierius committed:
[WTF] Define WTF_MAKE_CONFIGURABLE_ALLOCATED for container types
https://bugs.webkit.org/show_bug.cgi?id=286711
rdar://143849704

Reviewed by Yusuke Suzuki.

This allows us to specify that these types are allocated using the same
allocator as provided for their backing store, which is a more
comfortable default than "always use FastMalloc no matter what the
caller tells us they want".

If in the future someone does want to use separate allocators for the
two, we can add another template parameter to configure the two
separately as needed.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/Bag.h:
* Source/WTF/wtf/CheckedRef.h:
(WTF::CanMakeCheckedPtr::~CanMakeCheckedPtr):
(WTF::CanMakeThreadSafeCheckedPtr::~CanMakeThreadSafeCheckedPtr):
* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/FastMalloc.h:
(WTF::FastMalloc::fastFree):
(WTF::FastCompactMalloc::fastFree):
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope): Deleted.
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope): Deleted.
(WTF::TryMallocReturnValue::TryMallocReturnValue): Deleted.
(WTF::TryMallocReturnValue::~TryMallocReturnValue): Deleted.
(WTF::TryMallocReturnValue::getValue): Deleted.
* Source/WTF/wtf/FixedVector.h:
* Source/WTF/wtf/MallocCommon.cpp: Added.
(WTF::ForbidMallocUseForCurrentThreadScope::ForbidMallocUseForCurrentThreadScope):
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope):
(WTF::DisableMallocRestrictionsForCurrentThreadScope::DisableMallocRestrictionsForCurrentThreadScope):
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope):
* Source/WTF/wtf/MallocCommon.h: Added.
(WTF::TryMallocReturnValue::TryMallocReturnValue):
(WTF::TryMallocReturnValue::~TryMallocReturnValue):
(WTF::TryMallocReturnValue::getValue):
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope):
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope):
* Source/WTF/wtf/MallocSpan.h:
* Source/WTF/wtf/SegmentedVector.h:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::makeUnique):
(WTF::makeUniqueWithoutRefCountedCheck):
* Source/WTF/wtf/UniqueRef.h:
(WTF::makeUniqueRefWithoutRefCountedCheck):
(WTF::makeUniqueRef):
* Source/WTF/wtf/Vector.h:
* Source/WTF/wtf/text/WTFString.h:
* Source/bmalloc/bmalloc/IsoHeap.h:
* Source/bmalloc/bmalloc/IsoHeapInlines.h:
* Source/bmalloc/bmalloc/TZoneHeap.h:
* Source/bmalloc/bmalloc/TZoneHeapInlines.h:

Canonical link: https://commits.webkit.org/290626@main
